linera-core/src/client/mod.rs

```diff
@@ -2237,73 +2093,281 @@ impl<Env: Environment> ChainClient<Env> {
+                    // to have been already downloaded, so we will increase the
+                    // validators' trackers accordingly
+                    validator_trackers.downloaded_cert(ChainAndHeight {
+                        chain_id: *sender_chain_id,
+                        height: *height,
+                    });
+                }
             }
-            Err(error) => {
-                return Err(error.into());
+            remote_heights.retain(|h| *h >= local_next);
+        }
+
+        Ok(remote_heights)
+    }
+
+    /// Retries cross chain requests on the chains which may have been processed on
+    /// another task without the messages being correctly handled.
+    async fn retry_pending_cross_chain_requests(
+        &self,
+        nodes: &[RemoteNode<Env::ValidatorNode>],
+        other_sender_chains: Vec<ChainId>,
+    ) {
+        let stream = FuturesUnordered::from_iter(other_sender_chains.into_iter().map(|chain_id| {
+            let local_node = self.client.local_node.clone();
+            async move {
+                if let Err(error) = match local_node
+                    .retry_pending_cross_chain_requests(chain_id)
+                    .await
+                {
+                    Ok(()) => Ok(()),
+                    Err(LocalNodeError::BlobsNotFound(blob_ids)) => {
+                        if let Err(error) = self
+                            .client
+                            .update_local_node_with_blobs_from(blob_ids.clone(), nodes)
+                            .await
+                        {
+                            error!(
+                                "Error while attempting to download blobs during retrying outgoing \
+                                messages: {blob_ids:?}: {error}"
+                            );
+                        }
+                        local_node
+                            .retry_pending_cross_chain_requests(chain_id)
+                            .await
+                    }
+                    err => err,
+                } {
+                    error!("Failed to retry outgoing messages from {chain_id}: {error}");
+                }
             }
+        }));
+        stream.for_each(future::ready).await;
+    }
+
+    /// Downloads the certificates for blocks sending messages to this chain that we are
+    /// still missing.
+    async fn download_sender_certificates(
+        &self,
+        received_logs: Vec<(ValidatorPublicKey, Vec<ChainAndHeight>)>,
+        trackers: HashMap<ValidatorPublicKey, u64>,
+        nodes: Vec<RemoteNode<Env::ValidatorNode>>,
+        cancellation_token: Option<CancellationToken>,
+    ) -> Result<(), ChainClientError> {
+        let (received_logs, mut validator_trackers) = {
+            (
+                ReceivedLogs::from_received_result(received_logs.clone()),
+                ValidatorTrackers::new(received_logs, &trackers),
+            )
```

Comment:
Why not expect those types by the function right away? Instead of making the mapping at the very beginning